src/main/java/org/tikv/common/failsafe/CircuitBreakerMetricsImpl.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.tikv.common.failsafe;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CircuitBreakerMetricsImpl implements CircuitBreakerMetrics {
+  private static final Logger logger = LoggerFactory.getLogger(CircuitBreakerMetricsImpl.class);
+
+  private final int windowInMS;
+  private final List<MetricsListener> listeners;
+  private final AtomicReference<SingleWindowMetrics> currentMetrics;
+
+  public CircuitBreakerMetricsImpl(int windowInSeconds) {
+    this.windowInMS = windowInSeconds * 1000;
+    listeners = new ArrayList<>();
+    this.currentMetrics = new AtomicReference<>(new SingleWindowMetrics());
+  }
+
+  @Override
+  public void success() {
+    currentMetrics.get().success();
+    checkTimeout();
```

Comment:
@birdstorm @zz-jason should we use a `Timer` to check timeout?

Reply:
the current method is a little heavy because the `System.currentTimeMillis()`, would it be better to use `timer`?

Reply:
ok